### PR TITLE
[Impeller] improve blur performance for Android and iPad Pro.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Create a shader variant for blur that uses natively supported texture addressing modes, and a shader variant that emulates decal tile mode, removing the branch from the loop. I think this is legal? :)

Running some local experiments on wonderous on an iPad pro, the gaussian blur passes are currently taking  around 8-16ms. With this change applied, they're more consistently 4-8ms - roughly doubling the performance. While I didn't see an improvement on iOS in my past investigation of this change, this was likely because we were always running below frame time + I wasn't using the frame debugger. I'll investigate this separately, but I believe we'll also see a performance improvment on iPhones - and on Android as well.

While this does increase the number of shaders, its only by 1 - and for the most expensive filter we have Its probably worth it. Note that even 2xing performance, we still need to find a way to make some more gains to stay within budget for the iPad, probably at least one more doubling.

https://github.com/flutter/engine/pull/37927#issue-1465802725 contains details on the malioc output which confirm perfomance numbers for Android as well.

### Before

image

### After

image

### CODE REVIEW

File: impeller/entity/contents/filters/gaussian_blur_filter_contents.cc [hunk 1]
```diff
@@ -207,19 +205,48 @@ std::optional<Snapshot> DirectionalGaussianBlurFilterContents::RenderFilter(
...
+        source_descriptor.width_address_mode = SamplerAddressMode::kMirror;
+        source_descriptor.height_address_mode = SamplerAddressMode::kMirror;
+        break;
+      case Entity::TileMode::kRepeat:
+        cmd.pipeline = renderer.GetGaussianBlurPipeline(options);
+        input_descriptor.width_address_mode = SamplerAddressMode::kRepeat;
+        input_descriptor.height_address_mode = SamplerAddressMode::kRepeat;
+        source_descriptor.width_address_mode = SamplerAddressMode::kRepeat;
+        source_descriptor.height_address_mode = SamplerAddressMode::kRepeat;
+        break;
```
Comment (bdero) [inline impeller/entity/contents/filters/gaussian_blur_filter_contents.cc:239]: Oh yeah, we should totally be doing this! This has come up a couple of times before when the blur was under pressure.

**Review by bdero — APPROVED** [review]
LGTM assuming all of the TileModes still behave properly in `Play/EntityTest.GaussianBlurFilter/Metal`.

(When checking, note that some of the rarely used "BlurStyles" have been known broken for a while due to the blur optimizations, we need to get around to fixing it)

For future onlookers: Note that times in frame captures are extremely inaccurate and can very up to 5x in my experience. However, they're loosely precise, and so by normalizing results against the length of items in the capture which are expected to not be changed, loose performance comparisons can be made.
In the "before" screenshot, everything happens at ~half speed, and so at first glance, it appears that the first blur pass is 4x speed in the "after" screenshot, but it's actually closer to 2x. Another interesting observation is that the second blur pass is actually relatively slower in the "after" screenshot!

File: impeller/entity/contents/filters/gaussian_blur_filter_contents.cc [hunk 2]
```diff
@@ -281,7 +308,7 @@ std::optional<Rect> DirectionalGaussianBlurFilterContents::GetFilterCoverage(
 
   auto transform = inputs[0]->GetTransform(entity) * effect_transform;
   auto transformed_blur_vector =
-      transform.TransformDirection(blur_direction_ * Radius{blur_sigma_}.radius)
+      transform.TransformDirection(blur_direction_* Radius{blur_sigma_}.radius)
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/entity/contents/filters/gaussian_blur_filter_contents.cc:311]: Yup :(

**Comment by jonahwilliams** [comment]
I have visually verified the examples and they seem identical :)

**Comment by jonahwilliams** [comment]
Results:

https://flutter-flutter-perf.skia.org/e/?begin=1675291440&end=1675351360&keys=X66a0164e9e5c69d8a87c355849a6ffd9&selected=commit%3D33152%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253DiPhone_11%252Cdevice_version%253Dnone%252Chost_type%253Dmac%252Csub_result%253Daverage_frame_rasterizer_time_millis%252Ctest%253Dbackdrop_filter_perf_impeller_ios__timeline_summary%252C&xbaroffset=33152

**Pull request merged**